File: SkyrmDyn_Cpp/Plotting/setup_povray_file.py
import os

def setup_povray_file():
    input_file = open( "./SkyrmDyn_Cpp/inputs_config", "r" )
    file = input_file.readlines()
    for line in file:
        if line.find( "lattice_size = " ) != -1:
            sub_line = line[ 16:len(line)-2 ]
            lattice_size = sub_line.split()
            lattice_size[0] = float( lattice_size[0][ :len( lattice_size[0] )-1 ] )
            lattice_size[1] = float( lattice_size[1][ :len( lattice_size[1] )-1 ] )
            lattice_size[2] = float( lattice_size[2] )
            break
    input_file.close()
            
    look_at = [ 0.5*( lattice_size[0] + 1 ), 0.5*( lattice_size[1] + 1 ), 0.5*( lattice_size[2] + 1 ) ]
    position = [ 0.5*( lattice_size[0] + 1 ), 0.5*( lattice_size[1] + 1 ), 30*lattice_size[2] ]
    light = [ -40*lattice_size[0], -60*lattice_size[1], 100*lattice_size[2] ]
    
    camera = "camera { location < " + str( position[0] ) + ", " + str( position[1] ) + ", " + str( position[2] ) + " > look_at < " + str( look_at[0] ) + ", " + str( look_at[1] ) + ", " + str( look_at[2] ) + " > }\n"
    light_source = "light_source { < " + str( light[0] ) + ", " + str( light[1] ) + ", " + str( light[2] ) + " > color White }\n"
    
    povray_reader = open( "./SkyrmDyn_Cpp/Plotting/plot_spins.pov", "r" )
    povray_file = povray_reader.readlines()
    povray_reader.close()
    
    i = 0
    cpt = 0
    for line in povray_file:
        if cpt >= 2:
            break
        else:
            if line.find( "camera {" ) != -1:
                povray_file[i] = camera
                cpt += 1
            elif line.find( "light_source {" ) != -1:
                povray_file[i] = light_source
                cpt += 1
        i += 1
    
    povray_writer = open( "./SkyrmDyn_Cpp/Plotting/plot_spins_copy.pov", "w" )
    povray_writer.writelines( povray_file )
    povray_writer.close()
    
    os.remove( "./SkyrmDyn_Cpp/Plotting/plot_spins.pov" )
    os.rename( "./SkyrmDyn_Cpp/Plotting/plot_spins_copy.pov", "./SkyrmDyn_Cpp/Plotting/plot_spins.pov" )

if __name__ == "__main__":
    setup_povray_file()

File: SkyrmDyn_Cpp/Plotting/setup_povray_file.cpp
#include "Objects/Plot/plot.h"

#include "Plotting/setup_povray_file.h"

#include "Error_messages/error_messages.h"

#include "Inputs/Readers/inputs_reader.h"

#include "setup_povray_file.h"

#include "Maths/Objects/Array/Array_int/array_int.h"

#include "Inputs/Readers/inputs_reader.h"

#include "File_management/file_management.h"

#include <string>
#include <fstream>
#include <iostream>

#include <cmath>

void Plot::setup_povray_file() {
    std::string cmd = "\"\"" + python + "\" \"" + code + "SkyrmDyn_Cpp\\Plotting\\setup_povray_file.py\"\"";
    std::system( cmd.c_str() );
    /*
    set_camera();
    set_light_source();
    */
}

void move_pov_and_include_spin_file( std::string filename ) {
    move_file( "plot_spins.pov", "./SkyrmDyn_Cpp/Plotting/", "./Outputs/Povray_data_files/" );
    replace_line_in_file( "./Outputs/Povray_data_files/plot_spins.pov", "#include \"input_file.dat\"", "#include \"" + filename + "\"" );
}

void set_camera() {
    Array_int lattice_size = get_array_int( "lattice_size" );
    float x = lattice_size.get_x();
    float y = lattice_size.get_y();
    float z = lattice_size.get_z();

    float look_x = 0.5*x;
    float look_y = 0.5*y;
    float look_z = 0.5*z;

    x = -0.25*x;
    y = -0.4*y;
    z = 2*z;

    std::string old_line;
    std::ifstream read( "./SkyrmDyn_Cpp/Plotting/plot_spins.pov" );
    bool success = false;
    if( read ) {
        // Read the file line by line
        while( std::getline( read, old_line ) ) {
            // If the name of the line containing the parameter is found
            if( old_line.find( "camera {" ) != std::string::npos ) {
                success = true;
                break;
            }
        }
        if( !success ) {
            throw( OperationError( "finding", "old_line" ) );
        }
    }
    else {
        throw( DeclarationError( "\"read\"" ) );
    }

    std::string new_line = "camera { location < " + std::to_string(x) + ", "
                                                  + std::to_string(y) + ", "
                                                  + std::to_string(z) + " > look_at < "
                                                  + std::to_string(look_x) + ", "
                                                  + std::to_string(look_y) + ", "
                                                  + std::to_string(look_z) + " > }";

    replace_line_in_file( old_line, new_line, "./SkyrmDyn_Cpp/Plotting/plot_spins.pov" );
}

void set_light_source() {
    Array_int lattice_size = get_array_int( "lattice_size" );
    int l=0;
    for( int i=0; i<lattice_size.get_dimension(); i++ ) {
        l += lattice_size.get_component( i );
    }
    float x = -40*l;
    float y = -60*l;
    float z = 100*l;

    std::string old_line;
    std::ifstream read( "./SkyrmDyn_Cpp/Plotting/plot_spins.pov" );
    bool success = false;
    if( read ) {
        // Read the file line by line
        while( std::getline( read, old_line ) ) {
            // If the name of the line containing the parameter is found
            if( old_line.find( "light_source {" ) != std::string::npos ) {
                success = true;
                break;
            }
        }
        if( !success ) {
            throw( OperationError( "finding", "old_line" ) );
        }
    }
    else {
        throw( DeclarationError( "\"read\"" ) );
    }

    std::string new_line = "light_source { location < " + std::to_string(x) + ", "
                                                        + std::to_string(y) + ", "
                                                        + std::to_string(z) + " > color White }";

    replace_line_in_file( old_line, new_line, "./SkyrmDyn_Cpp/Plotting/plot_spins.pov" );
}

void write_nb_povray_files() {
    double llg_time = get_double( "llg_time" );
    double llg_time_step = get_double( "llg_time_step" );

    std::string filename = "./SkyrmDyn_Cpp/Plotting/plot_spins.pov";
    std::string new_line = "#declare nb_files = " + std::to_string( (int) round( llg_time / llg_time_step ) ) + ";";
    std::string line_to_find =  "#declare nb_files = ";
    std::ifstream read( filename );
    std::ofstream write( filename );

    std::string line;
    std::string tmp;
    if( read && write ) {
        // Read the file line by line
        while( std::getline( read, line ) ) {
            // If the name of the line containing the parameter is found
            if( line.find( line_to_find ) != std::string::npos ) {
                write << new_line;
                break;
            }
            tmp = line;
            write << line + "\n";
        }
    }
    else {
        throw( DeclarationError( "\"read\" and/or \"write\"" ) );
    }
}


